test/extended/networking/ipsec.go

```diff
@@ -0,0 +1,504 @@
+		// at node's primary interface. Based on the IPsec configuration on the cluster, the captured packet on the host interface
+		// must be either geneve or esp packet type.
+		type testNodeConfig struct {
+			pingPod    *corev1.Pod
+			tcpdumpPod *corev1.Pod
+			hostInf    string
+			nodeName   string
+		}
+		type testConfig struct {
+			ipsecConfig   *v1.IPsecConfig
+			srcNodeConfig *testNodeConfig
+			dstNodeConfig *testNodeConfig
+		}
+		// The config object contains test configuration that can be leveraged by each ipsec test.
+		var config *testConfig
+		// This function helps to generate ping traffic from src pod to dst pod and at the same captures its
+		// node traffic on both src and dst node.
+		pingAndCheckNodeTraffic := func(src, dst *testNodeConfig, ipsecTraffic bool) error {
+			tcpDumpSync := errgroup.Group{}
+			pingSync := errgroup.Group{}
+			// use tcpdump pod's ip address it's a node ip address because it's a hostnetworked pod.
+			srcNodeTrafficFilter := fmt.Sprintf(tcpdumpGeneveFilter, src.tcpdumpPod.Status.PodIP, dst.tcpdumpPod.Status.PodIP)
+			dstNodeTrafficFilter := fmt.Sprintf(tcpdumpGeneveFilter, dst.tcpdumpPod.Status.PodIP, src.tcpdumpPod.Status.PodIP)
+			if ipsecTraffic {
+				srcNodeTrafficFilter = fmt.Sprintf(tcpdumpESPFilter, src.tcpdumpPod.Status.PodIP, dst.tcpdumpPod.Status.PodIP)
+				dstNodeTrafficFilter = fmt.Sprintf(tcpdumpESPFilter, dst.tcpdumpPod.Status.PodIP, src.tcpdumpPod.Status.PodIP)
+			}
+			checkSrcNodeTraffic := func(src *testNodeConfig) error {
+				_, err := oc.AsAdmin().Run("exec").Args(src.tcpdumpPod.Name, "-n", src.tcpdumpPod.Namespace, "--",
+					"timeout", "10", "tcpdump", "-i", src.hostInf, "-c", "1", "-v", "--direction=out", srcNodeTrafficFilter).Output()
+				return err
+			}
+			checkDstNodeTraffic := func(dst *testNodeConfig) error {
+				_, err := oc.AsAdmin().Run("exec").Args(dst.tcpdumpPod.Name, "-n", dst.tcpdumpPod.Namespace, "--",
+					"timeout", "10", "tcpdump", "-i", dst.hostInf, "-c", "1", "-v", "--direction=out", dstNodeTrafficFilter).Output()
+				return err
+			}
+			pingTestFromPod := func(src, dst *testNodeConfig) error {
+				_, err := oc.AsAdmin().Run("exec").Args(src.pingPod.Name, "-n", src.pingPod.Namespace, "--",
+					"ping", "-c", "3", dst.pingPod.Status.PodIP).Output()
+				return err
+			}
+			tcpDumpSync.Go(func() error {
+				err := checkSrcNodeTraffic(src)
+				if err != nil {
+					return fmt.Errorf("error capturing traffic on the source node: %v", err)
+				}
+				return nil
+			})
+			tcpDumpSync.Go(func() error {
+				err := checkDstNodeTraffic(dst)
+				if err != nil {
+					return fmt.Errorf("error capturing traffic on the dst node: %v", err)
+				}
+				return nil
+			})
+			pingSync.Go(func() error {
+				return pingTestFromPod(src, dst)
+			})
+			// Wait for both ping and tcpdump capture complete and check the results.
+			pingErr := pingSync.Wait()
```

Comment:
- Does the tcpdump check return error when it captures 0 packets?
- Wouldn't it be good practice to check this error anyway?

Reply:
yes, it waits for pkt count to reach 1, otherwise it would wait until 10s and then throws timeout error.
its always good to check every error, will add the check for pingErr too :)

Reply:
added check for pingErr too,